### PR TITLE
prov/efa: Handle FI_ADDR_NOTAVAIL in rxr_ep_get_peer and add assertion against returned peer

### PR DESCRIPTION
**Pull request by zhngaj**
12ad455 Handle FI_ADDR_NOTAVAIL in rxr_ep_get_peer
25ddc04e9c6478c60b6d5cb9b2c08b98094aa720 Add assertion for peer structure
5a840a7d84816e83530e46b996f036cfcb69f998 Do not call rxr_ep_get_peer for local read

### CODE REVIEW

File: prov/efa/src/efa.h [hunk 1]
```diff
@@ -558,6 +558,10 @@ struct rdm_peer *rxr_ep_get_peer(struct rxr_ep *ep, fi_addr_t addr)
 {
 	struct util_av_entry *util_av_entry;
 	struct efa_av_entry *av_entry;
+
+	if (OFI_UNLIKELY(addr == FI_ADDR_NOTAVAIL))
```
Comment (rajachan) [inline prov/efa/src/efa.h:562]: Might as well add the assert inside here instead of returning NULL. Trying to get the peer object without an address is always a programming error and points to missing logic at the caller. Also the next person to call `rxr_ep_get_peer()` doesn't have to remember to add an assert :)

Comment (zhngaj) [reply prov/efa/src/efa.h:562]: assert() just takes effect in debug mode. I think it's still needed to return NULL here to avoid segfault due to the call to ofi_bufpool_get_ibuf.

File: prov/efa/src/rxr/rxr_read.c [hunk 2]
```diff
@@ -511,8 +513,10 @@ int rxr_read_post(struct rxr_ep *ep, struct rxr_read_entry *read_entry)
 
 	peer = rxr_ep_get_peer(ep, read_entry->addr);
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_read.c:514]: Let's move this call to inside the `if` condition.

Comment (zhngaj) [reply prov/efa/src/rxr/rxr_read.c]: Sure. Pushed new rev.

File: prov/efa/src/rxr/rxr_read.c [hunk 3]
```diff
@@ -509,10 +511,11 @@ int rxr_read_post(struct rxr_ep *ep, struct rxr_read_entry *read_entry)
 			return ret;
 	}
 
-	peer = rxr_ep_get_peer(ep, read_entry->addr);
-
-	if (read_entry->lower_ep_type == SHM_EP)
+	if (read_entry->lower_ep_type == SHM_EP) {
+		peer = rxr_ep_get_peer(ep, read_entry->addr);
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_read.c:515]: Can you make this change a separate commit because this change actually fixed the bug?

Comment (zhngaj) [reply prov/efa/src/rxr/rxr_read.c:515]: Actually, the bug is fixed by handling FI_ADDR_NOTAVAIL in rxr_ep_get_peer. Because we return NULL in time instead of calling the following function `ofi_bufpool_get_ibuf` (was hitting segfault here). 
That's why I made a bit change in my first revision based on our discussion, leaving the call to `rxr_ep_get_peer` outside of the if block.

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_read.c:515]: I see. But this change is different from other changes. In other places, we simply add `assert(peer)`, here we moved `rxr_ep_get_peer`. so

Comment (zhngaj) [reply prov/efa/src/rxr/rxr_read.c:515]: Yeah. Just not quite sure how to describe this change, as it seems unnecessary, (which is another reason I did not make this change in first revision). Any idea?

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_read.c:515]: do not call rxr_ep_get_peer for local read, because local read does not have a peer

Comment (zhngaj) [reply prov/efa/src/rxr/rxr_read.c:515]: Thanks, done in new rev.

**Comment by zhngaj** [comment]
multi-node #2184 was not stable
```
server_cmd: /home/ec2-user/libfabric/fabtests/install/bin/fi_rdm -U -p "efa" -E
server_stdout: |
Waiting for message from client...
Data check OK
Received data from client: Hello from Client!
client_cmd: /home/ec2-user/libfabric/fabtests/install/bin/fi_rdm -U -p "efa" -E 172.31.25.74
client_stdout: |
Killed by signal 15.
```

**Comment by zhngaj** [comment]
bot:aws:retest

**Comment by zhngaj** [comment]
@shefty could you paste the error shown in jenins/pr-head, trying to check if it's a real issue caused by this PR.

**Review by wzamazon — APPROVED** [review]

**Comment by shefty** [comment]
Intel CI:
```
name:   fi_av_xfer -e rdm -p "udp;ofi_rxd"
10:27:27    result: Fail
10:27:27    time:   116
10:27:27    server_cmd: /home/build/ofi-Install/libfabric/ofi_libfabric/PR-6854/4/dbg/bin/fi_av_xfer -e rdm -p "udp;ofi_rxd"  -s n7
10:27:27    server_stdout: |
10:27:27      AV address removal: PASS
10:27:27      AV re-insertion address: fi_av_xfer: ./include/ofi_mem.h:443: ofi_bufpool_get_ibuf: Assertion `ofi_buf_region(buf)->use_cnt' failed.
```

problem is unrelated here and has been fixed upstream.

**Comment by zhngaj** [comment]
> Intel CI:
> 
> ```
> name:   fi_av_xfer -e rdm -p "udp;ofi_rxd"
> 10:27:27    result: Fail
> 10:27:27    time:   116
> 10:27:27    server_cmd: /home/build/ofi-Install/libfabric/ofi_libfabric/PR-6854/4/dbg/bin/fi_av_xfer -e rdm -p "udp;ofi_rxd"  -s n7
> 10:27:27    server_stdout: |
> 10:27:27      AV address removal: PASS
> 10:27:27      AV re-insertion address: fi_av_xfer: ./include/ofi_mem.h:443: ofi_bufpool_get_ibuf: Assertion `ofi_buf_region(buf)->use_cnt' failed.
> ```
> 
> problem is unrelated here and has been fixed upstream.

Thank you.
